quartz-core/src/main/java/org/quartz/CronExpression.java

```diff
@@ -1589,9 +1590,409 @@ protected void setCalendarHour(Calendar cal, int hour) {
+                            cl.set(Calendar.MONTH, mon - 1);
+                            cl.set(Calendar.YEAR, cl.get(Calendar.YEAR) - 1);
+                        } else {
+                            cl.set(Calendar.DAY_OF_MONTH, getLastDayOfMonth(mon - 1, cl.get(Calendar.YEAR)));
+                            cl.set(Calendar.MONTH, mon - 1 - 1);
+                        }
+                        continue;
+                    } else if (daysToAdd < 0 || dayShifted) {
+                        cl.set(Calendar.SECOND, 0);
+                        cl.set(Calendar.MINUTE, 0);
+                        cl.set(Calendar.HOUR_OF_DAY, 0);
+                        cl.set(Calendar.DAY_OF_MONTH, day);
+                        cl.set(Calendar.MONTH, mon - 1);
+                        // '- 1' here because we are NOT promoting the month
+                        continue;
+                    }
+                } else {
+                    int cDow = cl.get(Calendar.DAY_OF_WEEK); // current d-o-w
+                    int dow = (int) daysOfWeek.last(); // desired
+                    // d-o-w
+                    st = daysOfWeek.headSet(cDow, true);
+                    if (st != null && st.size() > 0) {
+                        dow = st.last();
+                    }
+
+                    int daysToAdd = 0;
+                    if (cDow > dow) {
+                        daysToAdd = dow - cDow;
+                    }
+                    if (cDow < dow) {
+                        daysToAdd = dow - cDow - 7;
+                    }
+
+//                    int lDay = getLastDayOfMonth(mon, cl.get(Calendar.YEAR));
+
+                    if (day + daysToAdd < 1) { // will we pass the end of
+                        // the month?
+                        cl.set(Calendar.SECOND, 59);
+                        cl.set(Calendar.MINUTE, 59);
+                        cl.set(Calendar.HOUR_OF_DAY, 23);
+                        if (mon == 1) {
+                            mon = 12;
+                            cl.set(Calendar.DAY_OF_MONTH, getLastDayOfMonth(mon, cl.get(Calendar.YEAR)));
+                            cl.set(Calendar.MONTH, mon - 1);
+                            cl.set(Calendar.YEAR, cl.get(Calendar.YEAR) - 1);
+                        } else {
+                            cl.set(Calendar.DAY_OF_MONTH, getLastDayOfMonth(mon - 1, cl.get(Calendar.YEAR)));
+                            cl.set(Calendar.MONTH, mon - 1 - 1);
+                        }
+                        continue;
+                    } else if (daysToAdd < 0) { // are we swithing days?
+                        cl.set(Calendar.SECOND, 59);
+                        cl.set(Calendar.MINUTE, 59);
+                        cl.set(Calendar.HOUR_OF_DAY, 23);
+                        cl.set(Calendar.DAY_OF_MONTH, day + daysToAdd);
+                        cl.set(Calendar.MONTH, mon - 1);
+                        // '- 1' because calendar is 0-based for this field,
```

Comment:
This code can be moved to a private method, will remove the redundancy very much.